pkg/ipmi/ipmi.go

```diff
@@ -173,3 +188,61 @@ func (i *IPMI) ShutoffWatchdog() error {
 
 	return nil
 }
+
+func (i *IPMI) setsysinfo(data *setSystemInfoReq, dataLen uint16) error {
+	req := &req{}
+	req.msg.cmd = _SET_SYSTEM_INFO_PARAMETERS
+	req.msg.netfn = _IPMI_NETFN_APP
+	req.msg.dataLen = dataLen
+	req.msg.data = unsafe.Pointer(data)
+
+	if _, err := i.sendrecv(req); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SetSystemFWVersion sets the provided system firmware version to BMC via IPMI.
+func (i *IPMI) SetSystemFWVersion(version string) error {
+	strlenMax := 62 // Set 62 Bytes (4 sets) as the maximal version string length
+	len := len(version)
+
+	if len == 0 {
+		return fmt.Errorf("Version length is 0")
+	} else if len > strlenMax {
+		len = strlenMax
+	}
+
+	var data setSystemInfoReq
+	var copied int
+	var dataLen uint16
+	data.paramSelector = _SYSTEM_FW_VERSION
+	data.setSelector = 0
+	index := 0
+	for len > 0 {
+		if data.setSelector == 0 { // the fisrt block of string data
+			data.strData[0] = _ASCII
+			data.strData[1] = byte(len)
+			copied = copy(data.strData[2:], version)
+			// dataLen needs to add the actual copied string data plus the first 2 bytes
+			dataLen = uint16(int(unsafe.Sizeof(data.paramSelector)) + int(unsafe.Sizeof(data.setSelector)) + copied + 2)
+		} else {
+			copied = copy(data.strData[:], version[index:])
+			dataLen = uint16(int(unsafe.Sizeof(data.paramSelector)) + int(unsafe.Sizeof(data.setSelector)) + copied)
+		}
+
+		index += copied
+		if err := i.setsysinfo(&data, dataLen); err != nil {
+			return err
+		}
+
+		len -= copied
+		data.setSelector++
+		for j := range data.strData { //reset to 0
```

Comment:
Is this needed?

Reply:
The ipmitool uses strncpy, when the length of the source is smaller than the destination buffer it will write additional null bytes to the destination buffer to ensure that a total of n bytes are written . golang copy will not do this so I still zero out the buffer before the next iteration.